### ISSUE TITLE
Calling `.cloned()` on an `Option<X>` gives a misleading error message about iterators.

### ISSUE THREAD

**Issue by ssbr** [rust-lang/rust#151147]
### Code

```Rust
pub fn example(x: Option<i32>) -> i32 {
    x.cloned().unwrap()
}
```

### Current output

```Shell
error[E0599]: no method named `cloned` found for enum `Option<i32>` in the current scope
 --> src/lib.rs:5:7
  |
5 |     x.cloned().unwrap()
  |       ^^^^^^ `Option<i32>` is not an iterator
  |
help: call `.into_iter()` first
  |
5 |     x.into_iter().cloned().unwrap()
  |       ++++++++++++
```

### Desired output

```Shell
error[E0599]: no method named `cloned` found for enum `Option<i32>` in the current scope
 --> src/lib.rs:5:7
  |
5 |     x.cloned().unwrap()
  |       ^^^^^^ this is already an `Option<i32>`
  |
help: delete the call to `cloned`
  |
5 |     x.unwrap()
```

### Rationale and extra context

Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b8c24e8e97ad348d3d803be38cf0c213

This happens to me a lot, embarrassingly. The problem is that I _thought_ I had an `Option<&X>`, but in fact I have an `Option<X>`. For example, the following snippet of code:

```rust
let mut found = None;
for x in y {
  if ... {
    found = Some(x);
  }
}
x.cloned()
```

In this snippet, we never _explicitly_ spell out the type of `found`. In one version of the code, it might be `Option<&i32>`, because `y` implements `IntoIterator<Item=&i32>`. For example, maybe `y` was a `&[i32]`.

But if I later change `y` to be a `Vec<i32>`, or otherwise somehow end up with a non-reference when I wasn't really thinking about it, I get this confusing error message about iterators!

### Other cases

```Rust

```

### Rust Version

```Shell
I used the playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b8c24e8e97ad348d3d803be38cf0c213
```
